cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -235,66 +247,63 @@ private DatasetMeta getFromMds(DatasetId instance) throws Exception {
    */
   void create(String namespaceId, String name, DatasetInstanceConfiguration props) throws Exception {
     NamespaceId namespace = ConversionHelpers.toNamespaceId(namespaceId);
+    DatasetId datasetId = ConversionHelpers.toDatasetInstanceId(namespaceId, name);
     Principal principal = authenticationContext.getPrincipal();
-    authorizationEnforcer.enforce(namespace, principal, Action.WRITE);
+    String ownerPrincipal = props.getOwnerPrincipal();
+
+    // need to enforce on the principal id if impersonation is involved
+    KerberosPrincipalId principalId = ownerPrincipal == null ? null : new KerberosPrincipalId(ownerPrincipal);
+    if (!namespace.equals(NamespaceId.SYSTEM) && principalId == null) {
+      // if dataset owner is not present, get the namespace impersonation principal
+      String namespacePrincipal = ownerAdmin.getOwnerPrincipal(namespace);
+      principalId = namespacePrincipal == null ? null : new KerberosPrincipalId(namespacePrincipal);
+    }
+    if (principalId != null) {
+      authorizationEnforcer.enforce(principalId, principal, Action.ADMIN);
+    }
+    authorizationEnforcer.enforce(datasetId, principal, Action.ADMIN);
```

Comment:
This check should happen using authorization context since the principal should be the request user to handle all cases. Can you file a JIRA for this?